Add support for creating new releases for a github repo in shipitscript

Eijebong opened the pull request.

Eijebong committed:
Add support for creating new releases for a github repo in shipitscript

Eijebong committed:
Don't enable auto partials for ios

jcristau commented on shipitscript/src/shipitscript/script.py:
```diff
@@ -42,20 +42,28 @@ def create_new_release_action(context):
+        log.info(f"The shippable revision found is {shippable_revision}")
+    elif repository_type == "github":
+        shippable_revision = cron_revision
+        log.info(f"Repository type is github, shipping `cron_revision` {shippable_revision} without checking")
```
Can you explain?

Eijebong replied:
When the repository type is hg, shipit checks if there's any important commit since the last release by parsing the push log. The main issue is that the push log of a github project won't contain any of the markers (`a=...`, `DONTBUILD`, `L10n bumps`) so I decided against even thinking about it. I thought I had put a TODO on this though 🤔 Maybe I never saved the file.

jcristau replied:
Don't we still need to avoid building twice from the same rev?

Eijebong replied:
I guess we could get the latest release on that branch if there's one, get the SHA for that and compare?

Eijebong replied:
Done now

Eijebong replied:
https://firefox-ci-tc.services.mozilla.com/tasks/D3ghBHPPQfeYQnTnObt0TA/runs/0/logs/public/logs/live_backing.log tested here where the only commit on the release branch was the version bump that had just happened

jcristau commented on shipitscript/src/shipitscript/script.py:
```diff
@@ -42,20 +42,28 @@ def create_new_release_action(context):
+        shippable_revision = cron_revision
+        log.info(f"Repository type is github, shipping `cron_revision` {shippable_revision} without checking")
+    else:
+        log.error(f"Unknown repository type: {repository_type}. Exiting...")
```
Shouldn't this be a malformed payload error?

Eijebong replied:
Yup, done

Eijebong committed:
Change sending an unknown repo type to a TaskVerificationError

Eijebong committed:
Add support for repo_url in shipitscript

Eijebong committed:
Mirror the hg way of getting shippable revisions for github

jcristau commented on shipitscript/src/shipitscript/script.py:
```diff
@@ -59,14 +60,14 @@ def create_new_release_action(context):
             return
         log.info(f"The shippable revision found is {shippable_revision}")
-    elif repository_type == "github":
+    elif "github.com" in repository_url:
```
Might want to use mozilla-repo-urls (as a followup is also fine)?

jcristau commented on shipitscript/src/shipitscript/shipitapi.py:
```diff
@@ -110,10 +110,12 @@ def get_disabled_products(self, headers={}):
         resp = None
         params = {"product": product, "branch": branch, "version": version, "build_number": build_number, "revision": revision, "partial_updates": "auto"}
+        if repository_url is not None:
+            params["repo_url"] = repository_url
```
This might fix #931, at least partly?

Eijebong replied:
Actually I can fix it completely I think. I'll do that

jcristau commented on shipitscript/src/shipitscript/data/create_new_release_task_schema.json:
```diff
@@ -36,7 +36,7 @@
                   "type": "string"
                 },
-                "repository_type": {
+                "repository_url": {
```
I have a bit of a concern here with using an arbitrary repo url coming from the payload. Can we either pull it from something that can be verified to be the same repo that the task is coming from, or use scopes somehow, to prevent creating releases across repos?

Eijebong replied:
I now extract the repo url from the metadata->source of the task as that gets validated early.

Eijebong committed:
Derive the repo URL from the task metadata

Instead of having a user controllable field for this, derive it from the
task metadata which is checked by scriptworker earlier.

Eijebong committed:
Pass repo_url to shipit for hg repos too

Fixes #931

Eijebong commented on shipitscript/src/shipitscript/script.py:
```diff
@@ -43,22 +45,47 @@ def create_new_release_action(context):
+            log.info(f"Last shipped revision is {last_shipped_revision}")
+
+            log.info("Determining if cron revision is shippable")
+            repository_url = f"https://github.com/{source_url.owner}/{source_url.repo}"
```
This should be above line 68

Eijebong committed:
Set repository_url to something even if there's no last shipped revision

Also add a test for this

jcristau commented on shipitscript/src/shipitscript/script.py:
```diff
@@ -43,22 +45,48 @@ def create_new_release_action(context):
+        log.info(f"Last shipped revision is {last_shipped_revision}")
+
+        log.info("Determining most recent shippable revision")
+        repository_url = f"https://hg.mozilla.org/{source_url.repo}"
```
That won't work if source_url is None.  Though I think you can assert that it is not?

jcristau replied:
Also isn't repository_url just source_url.url, both in the hgmo and github cases?

Eijebong replied:
> That won't work if source_url is None. Though I think you can assert that it is not?

Oops, good catch, adding a test for this. I don't think we should assert that it's there, scriptworker doesn't unless CoT gets verified which isn't a guarantee AFAIK.

> Also isn't repository_url just source_url.url, both in the hgmo and github cases?

Nope

```diff
  -     'https://hg.mozilla.org/releases/mozilla-beta',
  +     'https://hg.mozilla.org/releases/mozilla-beta/file/8a2c52b9488e5ae5a8dc8500d3430d5ebe3d9a85/taskcluster/kinds/maybe-release'
```

Eijebong committed:
Don't assume that the source URL is set on a task

We already have a condition above that allows it to be None but this
would have tried to get the repo attribute of None if the task source
was actually missing.

jcristau commented on shipitscript/setup.py:
```diff
@@ -27,5 +27,5 @@
     license="MPL2",
     install_requires=requirements,
-    classifiers=("Programming Language :: Python :: 3.6", "Programming Language :: Python :: 3.7"),
+    classifiers=["Programming Language :: Python :: 3.6", "Programming Language :: Python :: 3.7"],
```
drive-by: that's a bit out of date :)

Eijebong replied:
Hahaha, I fixed that when I made sure that I didn't break setup.py when I removed the scriptworker_client dependency in the middle of this patch, because it turns out that it wasn't working in the first place. I can fix the classifiers if you want though

jcristau commented on shipitscript/src/shipitscript/shipitapi.py:
```diff
@@ -110,13 +110,15 @@ def get_disabled_products(self, headers={}):
         resp = None
         params = {"product": product, "branch": branch, "version": version, "build_number": build_number, "revision": revision, "partial_updates": "auto"}
+        if repository_url is not None:
+            params["repo_url"] = repository_url.format()
```
question: why .format()?

Eijebong replied:
Leftover and python is like "nah I don't care that it's empty" T_T
I'll fix that

Eijebong committed:
Remove stray `.format()`
